diff --git a/data/keyboards/Generic.kl b/data/keyboards/Generic.kl
index 02e0268e86de..0b1a3b468f49 100644
--- a/data/keyboards/Generic.kl
+++ b/data/keyboards/Generic.kl
@@ -182,9 +182,9 @@ key 159   FORWARD
 key 160   MEDIA_CLOSE
 key 161   MEDIA_EJECT
 key 162   MEDIA_EJECT
-key 163   MEDIA_NEXT
+key 163   APP_SWITCH      #MEDIA_NEXT
 key 164   MEDIA_PLAY_PAUSE
-key 165   MEDIA_PREVIOUS
+key 165   HOME            #MEDIA_PREVIOUS
 key 166   MEDIA_STOP
 key 167   MEDIA_RECORD
 key 168   MEDIA_REWIND
