diff --git a/cmds/app_process/app_main.cpp b/cmds/app_process/app_main.cpp
index 4e41f2c1ac35..1d8dc3da2ca5 100644
--- a/cmds/app_process/app_main.cpp
+++ b/cmds/app_process/app_main.cpp
@@ -333,6 +333,11 @@ int main(int argc, char* const argv[])
     }
 
     if (zygote) {
+        property_set("bst.config.pm_ready", "0");
+        property_set("bst.config.top_package_name", "");
+        property_set("bst.config.top_activity_name", "");
+        property_set("bst.config.boot_completed", "0");
+        property_set("bst.config.screen_enabled", "0");
         runtime.start("com.android.internal.os.ZygoteInit", args, zygote);
     } else if (!className.empty()) {
         runtime.start("com.android.internal.os.RuntimeInit", args, zygote);
