From 6828b1c43d5ea63d739bc4a0051fbae23499997d Mon Sep 17 00:00:00 2001
From: mark-bst <zeqing.xu@bluestacks.com>
Date: Wed, 5 Aug 2026 01:28:11 +0800
Subject: [PATCH] [A16] Install readiness probe runtime

---
 bstsvcmsgrtest/Android.bp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/bstsvcmsgrtest/Android.bp b/bstsvcmsgrtest/Android.bp
index e8692ee3d7..d4aa2520cf 100644
--- a/bstsvcmsgrtest/Android.bp
+++ b/bstsvcmsgrtest/Android.bp
@@ -6,4 +6,5 @@ cc_binary {
         "liblog",
         "libutils",
     ],
+    required: ["logwrapper"],
 }
-- 
2.34.1

