From cb141933695d9afa9109867e8041b4b914139fa7 Mon Sep 17 00:00:00 2001
From: mark-bst <zeqing.xu@bluestacks.com>
Date: Mon, 3 Aug 2026 19:33:00 +0800
Subject: [PATCH] [A16] Preserve OneStore in-app purchase handling

---
 core/java/android/app/Instrumentation.java | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/core/java/android/app/Instrumentation.java b/core/java/android/app/Instrumentation.java
index 704832b57c87..bf5a4322e64f 100644
--- a/core/java/android/app/Instrumentation.java
+++ b/core/java/android/app/Instrumentation.java
@@ -309,6 +309,17 @@ public class Instrumentation {
     }
 
     private boolean fwdIntentToHost(Context context, Intent intent) {
+        if (Process.myUid() >= Process.FIRST_APPLICATION_UID
+                && "com.onestore.component.iap.ACTION_PURCHASE_RESERVATION".equals(
+                        intent.getAction())) {
+            final String packageName = context.getBasePackageName();
+            final BstFilterAppsManager filterApps = BstFilterAppsManager.getInstance();
+            if (packageName != null && filterApps != null
+                    && filterApps.isOneStorePay(packageName)) {
+                return false;
+            }
+        }
+
         if (!intent.getBooleanExtra("openOnHostBrowser", false)) {
             return false;
         }
-- 
2.34.1

