From 6b2d5b08763e054484090e3aefbdf962b90952bf Mon Sep 17 00:00:00 2001
From: mark-bst <zeqing.xu@bluestacks.com>
Date: Tue, 4 Aug 2026 14:52:49 +0800
Subject: [PATCH] [A16] Reset the BlueStacks IME listener on switch

---
 .../server/inputmethod/InputMethodManagerService.java       | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/services/core/java/com/android/server/inputmethod/InputMethodManagerService.java b/services/core/java/com/android/server/inputmethod/InputMethodManagerService.java
index 1c80689c8064..b9b5693544fb 100644
--- a/services/core/java/com/android/server/inputmethod/InputMethodManagerService.java
+++ b/services/core/java/com/android/server/inputmethod/InputMethodManagerService.java
@@ -109,6 +109,7 @@ import android.os.ResultReceiver;
 import android.os.ShellCallback;
 import android.os.ShellCommand;
 import android.os.SystemClock;
+import android.os.SystemProperties;
 import android.os.Trace;
 import android.os.UserHandle;
 import android.os.UserManager;
@@ -2942,7 +2943,12 @@ public final class InputMethodManagerService implements IInputMethodManagerImpl.
         }
         final long ident = Binder.clearCallingIdentity();
         try {
+            final String previousImeId = bindingController.getSelectedMethodId();
             setSelectedInputMethodAndSubtypeLocked(info, subtypeIndex, false, userId);
+            if ("com.android.inputmethod.latin/.LatinIME".equals(previousImeId)
+                    && !id.equals(previousImeId)) {
+                SystemProperties.set("bst.config.ime_listenerport", "0");
+            }
             // mCurMethodId should be updated after setSelectedInputMethodAndSubtypeLocked()
             // because mCurMethodId is stored as a history in
             // setSelectedInputMethodAndSubtypeLocked().
-- 
2.34.1

