From 11209f2cb5806454e59ee2670822a447d5f60005 Mon Sep 17 00:00:00 2001
From: mark-bst <zeqing.xu@bluestacks.com>
Date: Wed, 5 Aug 2026 01:59:25 +0800
Subject: [PATCH] [A16] Update dumpstate diagnostics

---
 frameworks/native | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/frameworks/native b/frameworks/native
index c57571d..a98023c 160000
--- a/frameworks/native
+++ b/frameworks/native
@@ -1 +1 @@
-Subproject commit c57571d5cd966c7b4b3612898be7b5a2c7c77965
+Subproject commit a98023c677684dda5ffe2733ff0aaaafd12facef
-- 
2.34.1

