From 58434edc7d07b3bdb1294d6b780b58e630be87ec Mon Sep 17 00:00:00 2001
From: mark-bst <zeqing.xu@bluestacks.com>
Date: Wed, 5 Aug 2026 01:40:56 +0800
Subject: [PATCH] [A16] Update camera frame rate fallback

---
 frameworks/av | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/frameworks/av b/frameworks/av
index 6828b1c..5293270 160000
--- a/frameworks/av
+++ b/frameworks/av
@@ -1 +1 @@
-Subproject commit 6828b1c43d5ea63d739bc4a0051fbae23499997d
+Subproject commit 52932704dcc721eacdc692bfe8990115ab8041d2
-- 
2.34.1

