From 27488f4954b2c2fc81c7715969cf1439ff05de0f Mon Sep 17 00:00:00 2001
From: mark-bst <zeqing.xu@bluestacks.com>
Date: Wed, 5 Aug 2026 03:43:01 +0800
Subject: [PATCH] [A16] Update VA-API driver discovery

---
 external/libva | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/external/libva b/external/libva
index c474e2c..313d3c0 160000
--- a/external/libva
+++ b/external/libva
@@ -1 +1 @@
-Subproject commit c474e2c752b20997902df1f1b500f0439a58b3a5
+Subproject commit 313d3c0dbf612914d17cd1a02927c72a78e87d81
-- 
2.34.1

