From fe09ab6b0e1c08c56bdc947dea5d689e735f6268 Mon Sep 17 00:00:00 2001
From: muzhong-bst <muzhong@bluestacks.com>
Date: Thu, 14 Sep 2023 14:39:08 +0800
Subject: [PATCH 016/224] [A13] Disable Keyguard

---
 .../android/systemui/keyguard/KeyguardViewMediator.java    | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java b/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
index 1b0356141432..c542b3cc7741 100644
--- a/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
+++ b/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
@@ -2015,16 +2015,21 @@ public class KeyguardViewMediator implements CoreStartable, Dumpable,
      * @see #handleShow
      */
     private void showLocked(Bundle options) {
-        Trace.beginSection("KeyguardViewMediator#showLocked acquiring mShowKeyguardWakeLock");
+        // Trace.beginSection("KeyguardViewMediator#showLocked acquiring mShowKeyguardWakeLock");
         if (DEBUG) Log.d(TAG, "showLocked");
+        return;
         // ensure we stay awake until we are finished displaying the keyguard
+        /*
         mShowKeyguardWakeLock.acquire();
         Message msg = mHandler.obtainMessage(SHOW, options);
+        */
         // Treat these messages with priority - This call can originate from #doKeyguardTimeout,
         // meaning the device should lock as soon as possible and not wait for other messages on
         // the thread to process first.
+        /*
         mHandler.sendMessageAtFrontOfQueue(msg);
         Trace.endSection();
+        */
     }
 
     /**
