diff --git a/libs/WindowManager/Shell/src/com/android/wm/shell/transition/Transitions.java b/libs/WindowManager/Shell/src/com/android/wm/shell/transition/Transitions.java
index 92c4fcce3e00..4b82da139a21 100644
--- a/libs/WindowManager/Shell/src/com/android/wm/shell/transition/Transitions.java
+++ b/libs/WindowManager/Shell/src/com/android/wm/shell/transition/Transitions.java
@@ -145,7 +145,8 @@ public class Transitions implements RemoteCallable<Transitions>,
             SystemProperties.getBoolean("persist.wm.debug.finish_shell_transition", false);
 
     /** Set to {@code true} to enable shell transitions. */
-    public static final boolean ENABLE_SHELL_TRANSITIONS = true;
+    // TEMP(R262) / BST bringup: disable shell transitions until BLAST commit fixed (stuck EXITING)
+    public static final boolean ENABLE_SHELL_TRANSITIONS = false;
     public static final boolean SHELL_TRANSITIONS_ROTATION =
             SystemProperties.getBoolean("persist.wm.debug.shell_transit_rotate", false);
 
@@ -384,13 +385,16 @@ public class Transitions implements RemoteCallable<Transitions>,
                 new SettingsObserver());
 
         // Register this transition handler with Core
-        if (unifyShellBinders()) {
-            mOrganizer.initializeDependencies(this);
-        } else {
-            try {
-                mOrganizer.registerTransitionPlayer(mPlayerImpl);
-            } catch (RuntimeException e) {
-                throw e;
+        // TEMP(R262b) / BST bringup: gate TransitionPlayer on ENABLE_SHELL_TRANSITIONS
+        if (ENABLE_SHELL_TRANSITIONS) {
+            if (unifyShellBinders()) {
+                mOrganizer.initializeDependencies(this);
+            } else {
+                try {
+                    mOrganizer.registerTransitionPlayer(mPlayerImpl);
+                } catch (RuntimeException e) {
+                    throw e;
+                }
             }
         }
         // Pre-load the instance.
